Notifier services

qooplmao opened the pull request:
I don't know if this is of any interest to you as it creates quite a big breaking change due to the total removal of the `Elao\ErrorNotifierBundle\Listener\Notifier` class. Pretty much all of the original code is still there, it's just been moved about a bit.

I basically just wanted to be able to post any errors/exceptions to a bug tracker using a guzzle request so I did a bit of messing about. My thinking was to be able to use services as notifiers rather than only have the email option. In the process of adding this I ended up refactoring a bunch of stuff like..
- Changing Listener\Notifier to Listener\ExceptionListener and only having the actual event listeners
- Moving the configuration into a specific class
- Moving all of the exception handling stuff into an exception handler
- Moving the email notifier into a single use service
- Adding a notification handler that checks for the repeat timeout and the calls all of the enabled notifier services

The notifier services can be adding using the `elao.error_notifier` tag with a notifier name set as an alias which you can then enable using the configuration.

I've tried to keep everything so that it would work the same way without any config changes, so the removal of the `Listener\Notifier` class should be the only BC but I could be wrong.

lunetics commented:
+1 for this great change to have custom notifiers!

Should be good for a new major version!

qooplmao commented:
If this were to be for a major release then it wouldn't be as necessary to make sure the configuration stayed the same.

For example it might make more sense to move the mailer config into a notifiers section that auto added to the `enabled_notifiers` if enabled like (also, maybe underscore rather than capitalise because, for some reason, it doesn't feel right)..

```yml
elao_error_notifier:
    handle404: true # default :  false
    handlePHPErrors: true # catch fatal erros and email them
    handlePHPWarnings: true # catch warnings and email them
    handleSilentErrors: false # don't catch error on method with an @
    ignoredClasses: []
    enabled_notifiers: []
    enabled: true
    notifiers:
        mailer:
            enabled: true
            mailer: your.mailer.id
            from: from@example
            to: to@exmaple or [to@example]
```

This way it would allow for the possibility of extra default notifiers with config being added without having to confuse things too much.

This is just me thinking out loud though as this bundle is actually nothing to do with me.

lunetics commented:
Yes, sounds better for me. E.g. you could get rid of mailer.enabled and just use enabled_notifiers to check if a notifier is enabled. So every custom notifier could / should have it's own config.

ollietb commented on Exception/FlattenException.php:
```diff
@@ -0,0 +1,49 @@
+    /** Symfony 2.3+ */
+    const DEBUG_FLATTEN     = 'Symfony\Component\Debug\Exception\FlattenException';
+    /** Symfony <2.3 (deprecated in 2.3) */
+    const CONFIG_FLATTEN    = 'Symfony\Component\Debug\Exception\FlattenException';
```
Is this supposed to be 'Symfony\Component\HttpKernel\Exception\FlattenException'?

qooplmao replied:
Yeah that would be correct. I thing I just copied and pasted the line above and forgot to update it.

ollietb commented on Notifier/NotifierCollection.php:
```diff
@@ -0,0 +1,64 @@
+     */
+    public function getAllDisabled()
+    {
+        return $this->notifiers['enabled'];
```
Is this correct? getAllDisabled() returns enabled notifiers?

qooplmao replied:
Again, a bit of a typo.

ollietb commented:
I've often thought about doing something similar myself to send errors to a Slack channel. At the moment I'm using IFTTT but it's kind of messy. This would be a good enhancement.

qooplmao commented:
I just corrected those typos and dropped the "enabled" after @lunetics pointing out that it was already covered by the "enabled_notifiers"

qooplmao commented:
I've just added a slack notifier and a `ignored404Paths` config option as I want to be able to add these to a project that I am doing. I'm not sure if this will all be suitable for a PR any more as a lot has changed but I'll leave it up for the time being.

lunetics commented on README.md:
```diff
@@ -59,14 +59,29 @@ Add in your `config_prod.yml` file, you don't need error notifier when you are i
+                           # ignored if `notifiers.mailer.from` is set
+    to: to@example.com # required if using "default_mailer" notifier
+                       # can be a string or an array or email addresses
+                       # ignored if `notifiers.mailer.to` is set
```
Shouldn't we ditch the elao_error_notifier.to and elao_error_notifier.from ? and set the default notifier to the mailer?

qooplmao replied:
I've only left them there for backward compatibility for the configuration at the moment.
I did mention it a little lower down in the readme.

-------

### How to send error via email?

Enter the email addresses you wish to mail from and to (this can either be a string from 1 recipient or an array for multiple) and add the notifier `default_mailer` to your `enabled_notifiers` list.

```yml
# app/config/config_prod.yml
elao_error_notifier:
    # from: from@example.com 
    # to: [ to@example.com ]
    # mailer configurations at root are for backward compatability
    # and may be dropped so it is best to avoid
    enabled_notifiers:
        - default_slack
    notifier:
        mailer:
            from: from@example.com
            to: [ to@example.com ]
    
```

lunetics commented:
Good work so far, gonna test that in a bit and maybe add an hipchat notifier

qooplmao commented:
Any thoughts on this, pro or con?

Tom32i commented on Listener/ExceptionListener.php:
```diff
@@ -0,0 +1,79 @@
+     */
+    public function onKernelRequest(GetResponseEvent $event)
+    {
+    //    $this->exceptionHandler->initializeHandler($event->getRequest());
```
Commented line?

benji07 commented:
:+1:  could you rebase this PR using the last release, and then i will review the code for typo & style

lunetics commented:
Will this be ready soon?

qooplmao commented:
To be honest I'd totally forgotten about this.
I'll try to sort it over the weekend unless I get a chance before.

qooplmao commented:
I attempted to rebase and what not but feel I may have massively cocked it up somehow.

I've updated the code with the latest updates.
I've then moved all of the decision making stuff into "deciders" that work like the notifiers collection/services.
If 1 "decider" says to ignore the error then it gets ignored otherwise it will notify.

I've added a bunch of tests so some bits work.
I've not used it in production yet so it may not be 100%.

To be honest I have just gone totally rogue so I apologise if it's a bag of balls.

lunetics commented:
I 'll look over it maybe i got some ideas too :)

qooplmao commented:
I added a "ignore command" decider because I was getting a bunch of errors in the "cache:warmup" in deployment. This may be a sign of other issues in my deployment or in this edited script. I guess we shall see.

benji07 commented:
I think this is almost ready to be merge. But i found some errors.

- You have extra files that you should remove (.gitmodules, vagrant, Vagrantfile.dist)
- You add some stuff inside the .gitignore that could go to your global gitignore configuration (.vagrant, coverage, VagrantFile, ...)
- why default notifiers are named default_mailer & default_slack ?
- Some part of your code is not PSR2 compliant
- you need to rebase this branch
- inside the documentation you add the configuration key enabled_notifiers but inside the code you use enabledNotifiers
- You define constant inside the DeciderInterface but you didn't use them.
- inside your mailer configuration you don't allow the user to define his own mailer service

BC Break : When i test this against one of my application, the notification was not sent (i didn't change the configuration so i don't have the enabledNotifiers key)



I'm not sur we need a slack notifier as you just send informations that could be sent using monolog.

qooplmao commented:
I start doing stuff with this and then (like with everything else not work related) work takes over my time leaving this forgotten. The last thing I was doing was setting up a bunch of integration tests for the original bundle so that I can make sure my version handles things the same way... and then work...

I'll bang through all of these points at some point, but with Christmas (and work, and life, and yada yada yada) I'm not sure when I will get a chance. With regards to the BC break I did that on purpose due to https://github.com/Elao/ErrorNotifierBundle/pull/41#issuecomment-74419052 but I will put it back when I sort all of this out.
